File: sources/scala/14_file_read/main.scala
// task 14 file_read — expected output: 484442112
// build: scalac -release 17 main.scala    run: scala Main
// Reads data.bin (104857600 bytes) from the working directory in 1 MiB chunks.

import java.io.FileInputStream

object Main {
  def main(args: Array[String]): Unit = {
    val in = new FileInputStream("data.bin")
    try {
      val buf = new Array[Byte](1024 * 1024)
      var total = 0L
      var n = in.read(buf)
      while (n > 0) {
        var i = 0
        while (i < n) {
          total += (buf(i) & 0xFF).toLong
          i += 1
        }
        n = in.read(buf)
      }
      println(total % 4294967296L)
    } finally {
      in.close()
    }
  }
}

File: sources/kotlin/14_file_read/main.kt
// task 14 file_read — expected output: 484442112
// build: kotlinc main.kt -include-runtime -d prog.jar    run: java -jar prog.jar    (jvm row only)
// note: java.io is not available on Kotlin/Native, so this file targets the jvm row; the native row would need platform.posix read(2). Reads data.bin from the working directory in 1 MiB chunks.

import java.io.FileInputStream

fun main() {
    val buf = ByteArray(1024 * 1024)
    var total = 0L
    FileInputStream("data.bin").use { input ->
        while (true) {
            val got = input.read(buf)
            if (got <= 0) break
            var i = 0
            while (i < got) {
                total += (buf[i].toInt() and 0xFF).toLong()
                i++
            }
        }
    }
    println(total % 4294967296L)
}
